src/pyhf/__init__.py

```diff
@@ -138,17 +133,44 @@ def set_backend(backend, custom_optimizer=None, precision=None):
     else:
         new_optimizer = optimize.scipy_optimizer()
 
-    optimizer_changed = bool(optimizer != new_optimizer)
-    # set new backend
-    tensorlib = backend
-    optimizer = new_optimizer
-    # trigger events
-    if tensorlib_changed:
-        events.trigger("tensorlib_changed")()
-    if optimizer_changed:
-        events.trigger("optimizer_changed")()
-    # set up any other globals for backend
-    tensorlib._setup()
+    if default:
+        global default_backend
+        global default_optimizer
+        # need to determine if the default tensorlib changed or the default optimizer changed for events
```

Comment:
noting for now -- but a problem here is that we can't re-run the `__init__` of many things with events trigger. Likely, a bunch of the other code will need to be refactored so that we can change the default like this. We might need to consider having the default backend being set up at `import` time instead of being configurable like this. Might be a more sane option, otherwise, we end up needing to account for a significant more logic than expected.

Reply:
hm I see, how would a user configure that ahead of time before `import pyhf`?

Reply:
One thing to ponder is whether we might want to just drop support for changing backcends.. and just requirre people to choose ahead of time (like matplotlib has backends etc)

Reply:
> drop support for changing backcends

👀

I think at the present state, it's very helpful for users to be able to change quickly (and helps us gain more trust with them). I would not be opposed in the long term, but at the moment... ehh.

@phinate -- an idea would be to do something more like what matplotlib does

`pyhf.use('numpy')` to set the default tensor backend at the very start after importing pyhf.